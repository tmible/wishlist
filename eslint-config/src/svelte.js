import svelte from "eslint-plugin-svelte";
import globals from "globals";

export default [
  ...svelte.configs["flat/recommended"],
  {
    languageOptions: {
      globals: {
        ...globals.browser,
        ...globals.node,
      },
      ecmaVersion: "latest",
      sourceType: "module",
      parserOptions: {
        extraFileExtensions: [ ".svelte" ],
      },
    },

    settings: {
      "import/parsers": {
        "svelte-eslint-parser": [ ".svelte" ],
        "espree": [ ".js" ],
      },
    },

    rules: {
      /* import */
      // svelte does tree shaking when compiling
      "import/no-namespace": "off",

      /* svelte */
      "svelte/infinite-reactive-loop": "error",
      "svelte/no-dom-manipulating": "error",
      "svelte/no-dupe-on-directives": "error",
      "svelte/no-dupe-use-directives": "error",
      "svelte/no-reactive-reassign": "error",
      "svelte/no-store-async": "error",
      "svelte/require-store-callbacks-use-set-param": "error",
      "svelte/block-lang": "error",
      "svelte/no-ignored-unsubscribe": "error",
      "svelte/no-immutable-reactive-statements": "error",
      "svelte/no-inline-styles": "error",
      "svelte/no-reactive-functions": "error",
      "svelte/no-reactive-literals": "error",
      "svelte/no-useless-mustaches": "error",
      "svelte/require-each-key": "error",
      "svelte/require-event-dispatcher-types": "error",
      "svelte/require-stores-init": "error",
      "svelte/valid-each-key": "error",
      "svelte/derived-has-same-inputs-outputs": "error",
      "svelte/first-attribute-linebreak": "error",
      "svelte/html-closing-bracket-new-line": "error",
      "svelte/html-closing-bracket-spacing": "error",
      "svelte/html-quotes": "error",
      "svelte/html-self-closing": [
        "error",
        {
          "void": "never",
          "normal": "never",
          "svg": "always",
          "math": "never",
          "component": "always",
          "svelte": "always",
        },
      ],
      "svelte/indent": [ "error", { ignoredNodes: [ "ConditionalExpression" ] } ],
      "svelte/max-attributes-per-line": [ "error", { multiline: 1, singleline: 100 } ],
      "svelte/mustache-spacing": "error",
      "svelte/no-extra-reactive-curlies": "error",
      "svelte/no-spaces-around-equal-signs-in-attribute": "error",
      "svelte/prefer-class-directive": "error",
      "svelte/prefer-style-directive": "error",
      "svelte/shorthand-attribute": "error",
      "svelte/sort-attributes": [
        "error",
        {
          order: [
            // `this` property.
            "this",
            // `bind:this` directive.
            "bind:this",
            // `id` attribute.
            "id",
            // `name` attribute.
            "name",
            // `slot` attribute.
            "slot",
            // `--style-props`
            { match: "/^--/u", sort: "ignore" },
            // `style` attribute, and `style:` directives.
            [ "style", "/^style:/u" ],
            // `class` attribute.
            "class",
            // `class:` directives.
            { match: "/^class:/u", sort: "ignore" },
            // other attributes.
            {
              match: [ "!/:/u", "!/^(?:this|id|name|style|class)$/u", "!/^--/u" ],
              sort: "ignore",
            },
            // `bind:` directives (other then `bind:this`), and `on:` directives.
            [ "/^bind:/u", "!bind:this", "/^on:/u" ],
            // `use:` directives.
            { match: "/^use:/u", sort: "ignore" },
            // `transition:` directive.
            { match: "/^transition:/u", sort: "ignore" },
            // `in:` directive.
            { match: "/^in:/u", sort: "ignore" },
            // `out:` directive.
            { match: "/^out:/u", sort: "ignore" },
            // `animate:` directive.
            { match: "/^animate:/u", sort: "ignore" },
            // `let:` directives.
            { match: "/^let:/u", sort: "ignore" },
          ],
        },
      ],
      "svelte/spaced-html-comment": "error",
      "svelte/no-inner-declarations": "error",
      "svelte/no-trailing-spaces": "error",
    },
  },
  {
    files: [ "**/*.svelte" ],
    rules: {
      // svelte components can import themselves
      "import/no-self-import": "off",
    },
  },
  {
    files: [ "**/**.test.js", "**/**.test.svelte.js", "**/__tests__/**" ],
    settings: { "import/ignore": [ "vitest", "@testing-library/svelte" ] },
  },
];
