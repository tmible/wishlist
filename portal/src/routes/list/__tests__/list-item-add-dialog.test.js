// @vitest-environment jsdom
import { cleanup, render } from '@testing-library/svelte';
import { afterEach, beforeEach, describe, expect, it, vi } from 'vitest';
import ListItemAddDialog from '../list-item-add-dialog.svelte';

let formMock = null;

vi.mock('../list-item-form.svelte', () => ({ default: (_, props) => formMock = props }));

describe('list item add dialog', () => {
  let modalPortal;

  beforeEach(() => {
    modalPortal = document.createElement('div');
    modalPortal.id = 'modal-portal';
    document.body.append(modalPortal);
  });

  afterEach(() => {
    formMock = null;
    modalPortal.remove();
    vi.clearAllMocks();
    cleanup();
  });

  it('should be displayed closed', () => {
    const { baseElement } = render(ListItemAddDialog, { open: false });
    expect(baseElement).toMatchSnapshot();
  });

  describe('opened', () => {
    let baseElement;
    const add = vi.fn();

    beforeEach(() => {
      ({ baseElement } = render(ListItemAddDialog, { open: true, add }));
    });

    it('should be displayed', () => {
      expect(baseElement).toMatchSnapshot();
    });

    describe('on form cancel', () => {
      beforeEach(() => {
        formMock.cancel();
      });

      it('should close', () => {
        expect(baseElement).toMatchSnapshot();
      });

      it('should not dispatch event', () => {
        expect(add).not.toHaveBeenCalled();
      });
    });

    describe('on form success', () => {
      beforeEach(() => {
        formMock.success();
      });

      it('should close', () => {
        expect(baseElement).toMatchSnapshot();
      });

      it('should dispatch event', () => {
        expect(add).toHaveBeenCalled();
      });
    });
  });
});
