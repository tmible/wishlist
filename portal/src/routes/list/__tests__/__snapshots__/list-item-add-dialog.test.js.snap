// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`list item add dialog > opened > on form cancel > should close 1`] = `
<body
  style="margin-right: 8px; --scrollbar-width: 1024px; overflow: hidden; pointer-events: none;"
>
  <div
    id="modal-portal"
  >
    <!---->
    <!---->
    <!---->
    <!---->
    
    <!---->
     
    <!---->
    
    <!---->
    
    
    
  </div>
  <div>
    <!---->
    <!---->
    <!---->
    <!---->
    <button
      aria-controls="bits-13"
      aria-expanded="false"
      aria-haspopup="dialog"
      class="hidden"
      data-dialog-trigger=""
      data-state="closed"
      id="bits-11"
    >
      <!---->
    </button>
    
    <!---->
     
    <!---->
    
    <!---->
    
    
    
  </div>
</body>
`;

exports[`list item add dialog > opened > on form success > should close 1`] = `
<body
  style="margin-right: 8px; --scrollbar-width: 1024px; overflow: hidden; pointer-events: none;"
>
  <div
    id="modal-portal"
  >
    <!---->
    <!---->
    <!---->
    <!---->
    
    <!---->
     
    <!---->
    
    <!---->
    
    
    
  </div>
  <div>
    <!---->
    <!---->
    <!---->
    <!---->
    <button
      aria-controls="bits-27"
      aria-expanded="false"
      aria-haspopup="dialog"
      class="hidden"
      data-dialog-trigger=""
      data-state="closed"
      id="bits-25"
    >
      <!---->
    </button>
    
    <!---->
     
    <!---->
    
    <!---->
    
    
    
  </div>
</body>
`;

exports[`list item add dialog > opened > should be displayed 1`] = `
<body
  style="margin-right: 8px; --scrollbar-width: 1024px; overflow: hidden; pointer-events: none;"
>
  <div
    id="modal-portal"
  >
    <!---->
    <!---->
    <!---->
    <!---->
    <!---->
    <!---->
    <div
      class="modal-backdrop"
      data-dialog-overlay=""
      data-state="open"
      id="bits-5"
      style="pointer-events: auto;"
    >
      <!---->
    </div>
    
    
    
    <!---->
     
    <!---->
    <!---->
    <!---->
    <!---->
    <!---->
    <!---->
    <!---->
    <!---->
    <!---->
     
    <div
      aria-labelledby="bits-9"
      aria-modal="true"
      class="modal-box overflow-y-auto p-6"
      data-dialog-content=""
      data-state="open"
      id="bits-6"
      interactoutsidebehavior="close"
      role="dialog"
      style="pointer-events: auto; outline: undefined;"
      tabindex="-1"
    >
      <div
        class="flex items-start prose"
      >
        <!---->
        <div
          aria-level="2"
          class="mt-0 mr-auto"
          data-dialog-title=""
          data-state="open"
          id="bits-9"
          role="heading"
        >
          Добавить желание
          <!---->
        </div>
        
        <!---->
         
        <!---->
        <button
          class="cursor-pointer"
          data-dialog-close=""
          data-state="open"
          id="bits-10"
          tabindex="0"
        >
          <svg
            class="lucide-icon lucide lucide-x"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <!---->
            <path
              d="M18 6 6 18"
            >
              
            </path>
            
            <!---->
            <path
              d="m6 6 12 12"
            >
              
            </path>
            
            <!---->
            <!---->
            
            <!---->
          </svg>
          <!---->
        </button>
        
        <!---->
      </div>
       
      <!---->
      <!---->
    </div>
    
    
    
    
    
    
    
    <!---->
    
    
    
  </div>
  <div>
    <!---->
    <!---->
    <!---->
    <!---->
    <button
      aria-controls="bits-6"
      aria-expanded="true"
      aria-haspopup="dialog"
      class="hidden"
      data-dialog-trigger=""
      data-state="open"
      id="bits-4"
    >
      <!---->
    </button>
    
    <!---->
     
    <!---->
    
    <!---->
    
    
    
  </div>
</body>
`;

exports[`list item add dialog > should be displayed closed 1`] = `
<body>
  <div
    id="modal-portal"
  >
    <!---->
    <!---->
    <!---->
    <!---->
    
    <!---->
     
    <!---->
    
    <!---->
    
    
    
  </div>
  <div>
    <!---->
    <!---->
    <!---->
    <!---->
    <button
      aria-expanded="false"
      aria-haspopup="dialog"
      class="hidden"
      data-dialog-trigger=""
      data-state="closed"
      id="bits-1"
    >
      <!---->
    </button>
    
    <!---->
     
    <!---->
    
    <!---->
    
    
    
  </div>
</body>
`;
